### PR TITLE
fix: standalone workflows now fail on top-level cycles

### PR DESCRIPTION
**Pull request by yarinkos**
### Workflow toplevel resoruces size
There is no point for a workflow builder with zero size toplevel resources ( A bit illogically  ).

The following builder is very strange , but it demonstrates how wrong Workflow configuration could lead for a deadlock state (topLevelResources.size() == 0, cycle detection still does not implemented )
```
  initDependentResources(kubernetesClient);
  workflow = new WorkflowBuilder<WebPage>()
        .addDependentResource(configMapDR)
        .addDependentResource(deploymentDR).dependsOn(configMapDR)
        .addDependentResource(serviceDR).dependsOn(deploymentDR)
        .addDependentResource(ingressDR).dependsOn(serviceDR)
            .addDependentResource(configMapDR).dependsOn(serviceDR)
            .build();
```

For cases where the initializations aren't logical, this exception throw can comes in
Hope is following the guidelines and the project roadmap

### CODE REVIEW

**Comment by csviri** [comment]
Hi @yarinkos , thank you, this makes sense, just few points.
1. please run `mvn clean install` it will format the code, that is why the build is failing.
2. please add unit test for this

**Comment by csviri** [comment]
There is cycle detection for managed case: https://github.com/java-operator-sdk/java-operator-sdk/blob/077bbc2f224ee7fcc9da933aefd6c46aa51f95dd/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/workflow/ManagedWorkflowSupport.java#L59-L59

We could also properly do cycle detection to standalone mode. Still consider this as a good first step.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/workflow/DefaultWorkflow.java [hunk 1]
```diff
@@ -79,6 +79,9 @@ private Map<String, DependentResourceNode> toMap(Set<DependentResourceNode> node
       }
       map.put(node.getName(), node);
     }
+    if (topLevelResources.size() == 0) {
+            throw new IllegalStateException("Top level dependent resources set size must be bigger than 0 .");
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/workflow/DefaultWorkflow.java:83]: A more explicit message would be better as this doesn't really help the user fix the problem.

**Review by metacosm — CHANGES_REQUESTED** [review]

**Review by csviri — CHANGES_REQUESTED** [review]

**Review by csviri — APPROVED** [review]

**Review by metacosm — APPROVED** [review]
Thank you!

**Pull request merged**